### ISSUE TITLE
StringCopyPaste test phyically write to the clipboard causing flakyness

### ISSUE THREAD

**Issue by jmarolf** [dotnet/roslyn#61764]
In the [StringCopyPasteCommandHandlerTests](https://github.com/dotnet/roslyn/tree/main/src/EditorFeatures/CSharpTest/StringCopyPaste) it appears that the intent in that we call this [TestStringCopyPasteService](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/StringCopyPaste/TestStringCopyPasteService.cs), however locally I am seeing that these tests are instead calling [WpfStringCopyPasteService](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core.Wpf/StringCopyPaste/WpfStringCopyPasteService.cs). These means that while tests are running in VS you cannot interact with the clipboard or these tests will fail. It also means that they fail consistently if you run `Test.cmd` because we run these tests in separate processes in parallel but the OS clipboard is a singleton resource.

I suspect that this is because all of these test are marked as `WpfFact` that we are getting the actual `WpfStringCopyPasteService` instead of the `TestStringCopyPasteService`. @sharwell or @CyrusNajmabadi is this how we expect this set of tests to compose?

**Comment by CyrusNajmabadi** [comment]
Yes. It's testing actual copy/paste. :-)

**Comment by sharwell** [comment]
We likely need to use a named mutex to ensure tests that interact directly with the clipboard do not run in parallel, even if they are in separate processes.

> These means that while tests are running in VS you cannot interact with the clipboard or these tests will fail

This is a non-goal. It's similar to the requirement that you can't interact with the keyboard/mouse during integration tests.

**Comment by davidwengier** [comment]
If the tests interact with the "real world", perhaps moving them to the integration tests project would be the simplest thing to do?

**Comment by tmat** [comment]
> If the tests interact with the "real world", perhaps moving them to the integration tests project would be the simplest thing to do?

+1 We shouldn't be testing clipboard in unit tests.

**Comment by jmarolf** [comment]
Had a in-person conversation with @CyrusNajmabadi about this.

My understanding is that we are messing with the COM object that the editor places in the clipboard and so there is a need to actually verify that we are not corrupting it. While not perfect, the tests at least verify that against the current editor assemblies that we reference, we do not cause some sort of clip board corruption.

This does feel very "Integration Test"y  to me but there is the reality that our integration test leg is one of our longest and having these run on the regular CI machines greatly improves throughput for those developers who only test on CI machines.

My request is this:

- **Do not run these tests on `Test.cmd`**. It is my opinion that contributors should be able to do a "sanity check" of their branch and run a subset of our tests. Currently `Test.cmd` is what I use that for, and I would like us to not break that workflow.
- **Use xUnit to ensure that these tests are run sequentially per-machine instead of per-process**. I do not know exactly how to accomplish this, but given the infinite flexibility of xUnit I am making the assumption that this is possible. This ensures that if/when we refactor our CI infrastructure our tests properly explain their runtime requirements and we don't get surprise breaks.

@CyrusNajmabadi does this sound reasonable to you?

**Comment by sharwell** [comment]
> Do not run these tests on Test.cmd

This is fairly easy. We can use a trait on tests doing this, and exclude this trait for execution via Test.cmd (CI does not use Test.cmd, so we can put the exclusion directly in this file, or use other logic in Build.cmd to the same effect).

> Use xUnit to ensure that these tests are run sequentially per-machine instead of per-process

This is not so easy, especially for the way we use xunit. My recommendation is to use a new "machine mutex" with this form inside the test method itself:

```csharp
using var _ = MachineMutex.Take();
```

**Comment by jmarolf** [comment]
> My recommendation is to use a new "machine mutex" with this form inside the test method itself:

@sharwell so you are suggesting implementing some new helper type like this?

```csharp
internal static class SystemSemaphore
{
    private static Semaphore? _semaphore = null;
    private static bool _mutexWasCreated = false;

    public static IDisposable Take()
    {
        if (!Semaphore.TryOpenExisting(name: "Roslyn/GlobalsystemWideMutex", out _semaphore))
        {
            _semaphore = new Semaphore(initialCount: 1, maximumCount: 1, "Roslyn/GlobalsystemWideMutex", out _mutexWasCreated);

            if (!_mutexWasCreated)
            {
                throw new InvalidOperationException("Unable to create Semaphore");
            }
        }

        _ = _semaphore.WaitOne();
        return new SemaphoreDisposable(_semaphore);

    }

    private class SemaphoreDisposable : IDisposable
    {
        private readonly Semaphore _semaphore;

        public SemaphoreDisposable(Semaphore semaphore)
        {
            _semaphore = semaphore;
        }

        public void Dispose()
        {
            _semaphore.Release();
        }
    }
}
```

and consuming it from the tests?

**Comment by tmat** [comment]
I don't know how the editor clipboard code looks like but is it possible to mock the actual clipboard?

**Comment by CyrusNajmabadi** [comment]
The clipboard access service is acquired through MEF.  We currently have a windows-specific export.  We'll need one for cocoa as well.  Once we do, we can update the tests to run on both Windows and Mac test legs to validate they work properly.
